smoker/server/daemon.py

```diff
@@ -164,9 +162,33 @@ def run(self):
             signal.signal(signal.SIGTERM, self._shutdown)
         # API server loop now runs in a separate process and we don't want
         # to terminate to keep an instance of pluginmanager
-        while True:
-            time.sleep(1)
 
+        self._watchdog()
+
+    def _watchdog(self):
+        lg.debug("Starting the smoker watchdog")
+        setproctitle.setproctitle("smokerd watchdog")
+
+        while True:
+            if not self.server.is_alive():
+                lg.error("REST API server is dead")
+                self._restart_api_server()
+                lg.info("restarted the REST API server")
+
+            for plugin in self.pluginmgr.plugins.values():
+                if not plugin.is_alive():
+                    lg.error("Plugin %s is dead" % plugin.name)
+                    self.pluginmgr.restart_plugin(plugin.name)
+                    # need to restart the REST server so it has reference to
+                    # the new plugin instance
+                    self.server.terminate()
+                    self._restart_api_server()
+            time.sleep(10)
+
+    def _restart_api_server(self):
```

Comment:
Since it's only needed when the api server is still alive. It's required when restarting the API server after a restart of a plugin. This method is also used for restarting the server in case it dies/gets killed. In such case there's no need to terminate already dead process.